micro-rdk-nmea/src/messages/pgns.rs

```diff
@@ -106,3 +124,101 @@ pub struct GnssSatsInView {
     #[length_field = "sats_in_view"]
     satellites: Vec<Satellite>,
 }
+
+macro_rules! define_pgns {
+    ( $(($pgndef:ident, $enum:ident, $pgn:expr)),* ) => {
+        #[derive(Clone, Debug)]
+        pub enum MessageData {
+            $($enum($pgndef)),*,
+            Unsupported(UnparsedMessageData)
+        }
+
+        impl MessageData {
+            pub fn pgn(&self) -> u32 {
+                match self {
+                    $(Self::$enum(msg) => msg.pgn()),*,
+                    Self::Unsupported(unparsed) => unparsed.pgn()
+                }
+            }
+
+            pub fn from_bytes(pgn: u32, bytes: Vec<u8>) -> Result<Self, crate::parse_helpers::errors::NmeaParseError> {
+                Ok(match pgn {
+                    $($pgn => {
+                        let cursor = DataCursor::new(bytes);
+                        Self::$enum($pgndef::from_cursor(cursor)?)
+                    }),*,
+                    x => Self::Unsupported(UnparsedMessageData::from_bytes(bytes, x)?)
+                })
+            }
+
+            pub fn to_readings(self) -> Result<GenericReadingsResult, crate::parse_helpers::errors::NmeaParseError> {
+                match self {
+                    $(Self::$enum(msg) => msg.to_readings()),*,
+                    Self::Unsupported(msg) => msg.to_readings()
+                }
+            }
+        }
+    };
+}
+
+pub const MESSAGE_HEADER_OFFSET: usize = 32;
+
+define_pgns!(
+    (WaterDepth, Pgn128267, 128267),
+    (TemperatureExtendedRange, Pgn130316, 130316),
+    (GnssSatsInView, Pgn129540, 129540)
+);
+
+pub struct NmeaMessage {
+    metadata: NmeaMessageMetadata,
+    data: MessageData,
+}
+
+impl TryFrom<Vec<u8>> for NmeaMessage {
+    type Error = NmeaParseError;
+    fn try_from(mut value: Vec<u8>) -> Result<Self, Self::Error> {
+        let msg_data = value.split_off(MESSAGE_HEADER_OFFSET);
+        let metadata = NmeaMessageMetadata::try_from(value)?;
+        let data = MessageData::from_bytes(metadata.pgn(), msg_data)?;
```

Comment:
I feel like both these types (`NmeaMessageMetadata` and `MessageData`) should either start with `Nmea`, or not. Given that there is also NmeaMessage, maybe that breaks the tie in favor of `NmeaMessageData`?